src/services/review/ReviewRequestService.ts

```diff
@@ -0,0 +1,365 @@
+    // unable to find a RR with the tuple implies that said RR does not exist.
+    // This could happen when the user queries for an existing PR that is on github,
+    // but created prior to this feature rolling out.
+    if (!review) {
+      return new RequestNotFoundError()
+    }
+
+    // NOTE: We explicitly destructure as the raw data
+    // contains ALOT more than these fields, which we want to
+    // discard to lower retrieval times for FE
+    const { title, created_at } = await this.apiService.getPullRequest(
+      siteName,
+      pullRequestNumber
+    )
+
+    const changedItems = await this.compareDiff(userWithSiteSessionData)
+
+    return {
+      reviewUrl: review.reviewMeta.reviewLink,
+      title,
+      status: review.reviewStatus,
+      requestor: review.requestor.email || "",
+      reviewers: review.reviewers.map(({ email }) => email || ""),
+      reviewRequestedTime: new Date(created_at).getTime(),
+      changedItems,
+    }
+  }
+
+  /**
+   * Updates the review request with provided details.
+   * Note that the semantics for updating the review request description
+   * is as follows:
+   *
+   * 1. If the body is `undefined`, we **do not** update it.
+   * 2. If the body is `""` (an empty string), we update the\
+   * github pull request description to be empty too.
+   * 3. Otherwise, we just write through to github.
+   */
+  updateReviewRequest = async (
+    reviewRequest: ReviewRequest,
+    { title, description, reviewers }: RequestChangeInfo
+  ) => {
+    // Update db state with new reviewers
+    reviewRequest.reviewers = reviewers
+    const savePromise = reviewRequest.save()
+
+    const siteName = reviewRequest.site.name
+    const { pullRequestNumber } = reviewRequest.reviewMeta
+
+    // Update github state
+    const githubUpdatePromise = this.apiService.updatePullRequest(
+      siteName,
+      pullRequestNumber,
+      title,
+      description
+    )
+
+    await Promise.all([savePromise, githubUpdatePromise])
+  }
+
+  // NOTE: The semantics of our reviewing system is slightly different from github.
+  // The approval is tied to the request, rather than the user.
+  approveReviewRequest = async (reviewRequest: ReviewRequest) => {
+    reviewRequest.reviewStatus = "APPROVED"
+    await reviewRequest.save()
+  }
+
+  closeReviewRequest = async (reviewRequest: ReviewRequest) => {
+    const siteName = reviewRequest.site.name
+    const { pullRequestNumber } = reviewRequest.reviewMeta
+    await this.apiService.closeReviewRequest(siteName, pullRequestNumber)
+
+    reviewRequest.reviewStatus = "CLOSED"
+    await reviewRequest.save()
+  }
+
+  mergeReviewRequest = async (
+    reviewRequest: ReviewRequest
+  ): Promise<ReviewRequest | RequestNotFoundError> => {
+    const siteName = reviewRequest.site.name
+    const { pullRequestNumber } = reviewRequest.reviewMeta
+
+    await this.apiService.approvePullRequest(siteName, pullRequestNumber)
```

Comment:
Question: Why do we only approve here when merging and not when during the approve review request step?

Reply:
because our semantics are different from github - github's approval is _per user_, whereas ours is _per review_.

hence, this means that using github's approval is more annoying to work with (eg: base PR states are `open | closed` and we have to traverse the data structure + reconcile to see if the PR is approved)

Reply:
I see, that makes sense. Thanks for clarifying!